Spec edits for incremental delivery, Validation (#1223)
graphql/graphql-spec: spec/Section 5 -- Validation.md

}
```

### Defer And Stream Directives Are Used On Valid Root Field

** Formal Specification **

- For every {directive} in a document.
- Let {directiveName} be the name of {directive}.
- Let {mutationType} be the root Mutation type in {schema}.
- Let {subscriptionType} be the root Subscription type in {schema}.
- If {directiveName} is "defer" or "stream":
  - The parent type of {directive} must not be {mutationType} or

>     {subscriptionType}.

Reviewer: Let's split this up into a `@stream` and `@defer` rule:

```suggestion
**Formal Specification**

- For each `@stream` {directive} in the document:
  - Let {field} be the field to which the directive is applied.
  - The parent type of {field} must not be the {`mutation`} _root
    operation type_ or the {`subscription`} _root operation type_.
```

For defer it's on fragments so it's a little different. I'm not a big fan of this suggested edit, it's inspired by https://spec.graphql.org/draft/#sec-Fragment-Spread-Type-Existence but I feel like maybe we should branch on inline fragment vs named fragment spread.

```
- For each `@defer` {directive} in the document:
  - Let {fragmentType} be the target type of the fragment spread to which {directive} is applied.
  - {fragmentType} must not be the {`mutation`} _root
    operation type_ or the {`subscription`} _root operation type_.
```

Author: I rewrote this rule to use a CollectFields like approach instead. This will also catch cases where the interfaces are used on the root types. Also opened a PR to update it in GraphQL-JS https://github.com/graphql/graphql-js/pull/4761

**Explanatory Text**

The defer and stream directives are not allowed to be used on root fields of the

> mutation or subscription type.

Reviewer: ```suggestion
The `@defer` and `@stream` directives are not allowed to be used on root fields of the
mutation or subscription type.
```

For example, the following document will not pass validation because `@defer`
has been used on a root mutation field:

```raw graphql counter-example
mutation {
  ... @defer {
    mutationField
  }
}
```

### Defer And Stream Directives Are Used On Valid Operations

> ** Formal Specification **

Reviewer: ```suggestion
**Formal Specification**
```

**Explanatory Text**

The defer and stream directives can not be used to defer or stream data in
subscription operations. If these directives appear in a subscription operation
they must be disabled using the "if" argument. This rule will not permit any
defer or stream directives on a subscription operation that cannot be disabled
using the "if" argument.

For example, the following document will not pass validation because `@defer`
has been used in a subscription operation with no "if" argument defined:

```raw graphql counter-example
subscription sub {
  newMessage {
    ... @defer {
      body
    }
  }
}
```

### Defer And Stream Directive Labels Are Unique

** Formal Specification **

- Let {labelValues} be an empty set.
- For every {directive} in the document:
  - Let {directiveName} be the name of {directive}.
  - If {directiveName} is "defer" or "stream":
    - For every {argument} in {directive}:
      - Let {argumentName} be the name of {argument}.
      - Let {argumentValue} be the value passed to {argument}.
      - If {argumentName} is "label":
        - {argumentValue} must not be a variable.
        - {argumentValue} must not be present in {labelValues}.

>         - Append {argumentValue} to {labelValues}.

Reviewer 2: ```suggestion
        - Add {argumentValue} to {labelValues}.
```

  - {label} must not be present in {labelValues}.
  - Add {label} to {labelValues}.

**Explanatory Text**

The `@defer` and `@stream` directives each accept an argument `label`. This
label may be used by GraphQL clients to uniquely identify response payloads. If
a label is passed, it must not be a variable and it must be unique within all
other `@defer` and `@stream` directives in the document.

For example the following document is valid:

```graphql example
{
  dog {
    ...fragmentOne
    ...fragmentTwo @defer(label: "dogDefer")
  }
  pets @stream(label: "petStream") {
    name
  }
}

fragment fragmentOne on Dog {
  name
}

fragment fragmentTwo on Dog {
  owner {
    name
  }
}
```

The following document will not pass validation because the same label is used
in multiple `@defer` and `@stream` directives:

```raw graphql counter-example
{
  dog {
    ...fragmentOne @defer(label: "MyLabel")
  }
  pets @stream(label: "MyLabel") {
    name
  }
}

fragment fragmentOne on Dog {
  name
}
```

### Stream Directives Are Used On List Fields

**Formal Specification**

- For every `@stream` {directive} in the document:
- Let {adjacent} be the AST node {directive} affects.

> - {adjacent} must be a List type.

Reviewer: ```suggestion
- For every `@stream` {directive} in the document:
  - Let {adjacent} be the AST node {directive} affects.
  - Let {nullableFieldType} be unwrapped nullable type of {adjacent}.
  - {nullableFieldType} must be a List type.
```

**Explanatory Text**

GraphQL directive locations do not provide enough granularity to distinguish the
type of fields used in a GraphQL document. Since the stream directive is only
valid on list fields, an additional validation rule must be used to ensure it is
used correctly.

For example, the following document will only pass validation if `field` is

> defined as a List type in the associated schema.

Reviewer: ```suggestion
For example, the following document will only pass validation if `field`
contains a List type in the associated schema.
```
